lib/translate/status_translator.cc

```diff
@@ -0,0 +1,257 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "status_translator.h"
+
+#include <cstdio>
+
+#include "common.h"
+
+namespace translator {
+
+util::Optional<TranslatedScsiStatus> translateStatusCode(uint8_t sct,
+                                                         uint8_t sc) {
+  nvme_defs::StatusCodeType type = static_cast<nvme_defs::StatusCodeType>(sct);
+  switch (type) {
+    case nvme_defs::StatusCodeType::kGeneric:
+      return translateStatusCode(
+          static_cast<nvme_defs::GenericCommandStatusCode>(sc));
+    case nvme_defs::StatusCodeType::kCommandSpecific:
+      return translateStatusCode(
+          static_cast<nvme_defs::CommandSpecificStatusCode>(sc));
+    case nvme_defs::StatusCodeType::kMediaError:
+      return translateStatusCode(
+          static_cast<nvme_defs::MediaErrorStatusCode>(sc));
+    case nvme_defs::StatusCodeType::kPath:
+    case nvme_defs::StatusCodeType::kVendorSpecific:
+    default:
+      char debug_buffer[100];
+      sprintf(debug_buffer,
```

Comment:
can we make this into a general fn?

Reply:
This is an excellent idea. The `DebugLog` function I proposed [above](https://github.com/googleinterns/scsi2nvme/pull/62#discussion_r445142974) could be made variadic and call `snprintf` internally. Something more iostreams-style could also be done, but that'd probably be a lot more effort.

We should also double-check that `snprintf` is available in the kernel, and find an alternative approach if not.

Reply:
Yes, linux/include/linux/kernel.h includes snprintf.